sklearn/preprocessing/_polynomial.py

```diff
@@ -989,17 +1034,47 @@ def transform(self, X):
                     # This is equivalent to BSpline(.., extrapolate="periodic")
                     # for scipy>=1.0.0.
                     n = spl.t.size - spl.k - 1
-                    # Assign to new array to avoid inplace operation
-                    x = spl.t[spl.k] + (X[:, i] - spl.t[spl.k]) % (
-                        spl.t[n] - spl.t[spl.k]
-                    )
-                else:
-                    x = X[:, i]
+                    if spl.t[n] - spl.t[spl.k] > 0:
+                        # Assign to new array to avoid inplace operation
+                        x = spl.t[spl.k] + (X[:, feature_idx] - spl.t[spl.k]) % (
+                            spl.t[n] - spl.t[spl.k]
+                        )
+                    else:
+                        # This can happen if the column has a single non-nan
+                        # value. Treat as a constant feature.
+                        x = np.zeros_like(X[:, feature_idx])
+                else:  # self.extrapolation in ("continue", "error")
+                    x = X[:, feature_idx]
 
                 if use_sparse:
+                    # Copy the current column to avoid mutation of the user provided
+                    # input data when doing inplace operations.
+                    x = x.copy()
+                    # We replace the nan values in the input column by some
+                    # arbitrary, in-range, numerical value since
+                    # BSpline.design_matrix() would otherwise raise on any nan
+                    # value in its input. The spline encoded values in
+                    # the output of that function that correspond to missing
+                    # values in the original input will be replaced by 0.0
+                    # afterwards.
+                    #
+                    # Note that in the following we use np.nanmin(x) as the
+                    # input replacement to make sure that this code works even
+                    # when `extrapolation == "error"`. Any other choice of
+                    # in-range value would have worked work since the
+                    # corresponding values in the array are replaced by zeros.
+
+                    if nan_row_indices.size == x.size:
+                        # The column is all np.nan valued. Replace it by a
+                        # constant column with an arbitrary non-nan value
+                        # inside so that it is encoded as constant column.
+                        x[:] = 0
+                    else:
+                        x[nan_row_indices] = np.nanmin(x)
```

Comment:
```suggestion
                    if nan_row_indices.size == x.size:
                        # The column is all np.nan valued. Replace it by a
                        # constant column with an arbitrary non-nan value
                        # inside so that it is encoded as constant column.
                        x = np.zeros_like(x)  # avoid mutation of input data
                    elif nan_row_indices.shape[0] > 0:
                        x = x.copy()  # avoid mutation of input data
                        x[nan_row_indices] = np.nanmin(x)
```